### PR TITLE
Desktop Access Session with Toolbar

### PR DESCRIPTION
**Pull request by ibeckermayer**
This PR adds the core structure for the Desktop Access window with a toolbar. 

The toolbar itself has state based on the session settings (currently these settings `clipboard` and `recording` are hardcoded, but in the future will come from the backend).

Screen Shot 2021-09-13 at 10 25 24 AM
Screen Shot 2021-09-13 at 10 25 15 AM

The `...` menu also has a `Disconnect` option for the user to explicitly Disconnect the session

Screen Shot 2021-09-13 at 10 27 23 AM

The requirement of adding this `Disconnected` state made the overall state of this component more complex than what our typical `useAttempt` hook offers. To handle this, I added
```typescript
export type DesktopSessionAttempt = {
  status: Attempt['status'] | 'disconnected';
  statusText?: string;
};
```
and a "meta attempt" state
```typescript
const [attempt, setAttempt] = useState<DesktopSessionAttempt>({
    status: 'processing',
  });
```

This meta state tracks the logical combination of the websocket/tdpclient connection, the initial api call to fetch the Desktop (which will eventually evolve to or add another call to fetch a session with session settings like `clipboard` and `recording`), and whether or not the user has purposefully disconnected the session. See the storyboard for how those states look.

See the [**Test the system manually by**](https://github.com/gravitational/webapps/pull/385#issue-722861571) heading here for how to build/use Teleport with the Desktop Access beta feature.

### CODE REVIEW

File: packages/teleport/src/DesktopSession/TopBar.tsx
Comment (kimlisa) [inline packages/teleport/src/DesktopSession/TopBar.tsx:148]: hm, wondering why we couldn't just inline the styles? 
```
{ attempt.status === 'processing' && <Text style={...}> }
```

or we could also pass the attempt down to a styled comp and do our checking for states there?
```
const SomeName = styled.div`
    color: ${({ attempt }) => attempt.state === 'proccessing' ? ....};
`;
```

File: packages/teleport/src/DesktopSession/ActionMenu.tsx
Comment (ibeckermayer) [reply packages/teleport/src/DesktopSession/ActionMenu.tsx:33]: Confused by a few things here:

- Why can I pass `color` as an object field, but need to pass `fontSize` in the `style` object?
- How/where does the string `'text.secondary'` get interpreted to mean `props.theme.colors.text.secondary`? Or is that a misunderstanding of what the `StyledThemeProvider` is passing down to child objects?

File: packages/teleport/src/DesktopSession/ActionMenu.tsx
Comment (ibeckermayer) [reply packages/teleport/src/DesktopSession/ActionMenu.tsx:62]: Seems not. I also removed from `packages/teleport/src/Console/ActionBar/ActionBar.tsx` and `packages/teleport/src/Player/ActionBar/ActionBar.tsx`.

File: packages/teleport/src/DesktopSession/DesktopSession.tsx
Comment (ibeckermayer) [reply packages/teleport/src/DesktopSession/DesktopSession.tsx:45]: The reason its out here is that on a render event, we need to draw to the canvas, and the canvasRef necessarily lives here. My initial thought was that keeping all of the tdpClient event listeners in one place was the best design. 
```typescript
tdpClient.on('render', ({ bitmap, left, top }) => {
      const ctx = canvasRef.current.getContext('2d');
      ctx.drawImage(bitmap, left, top);
    });
```

But now that I'm thinking about it again, it would be preferable to keep all the `setXXX` state in the state hook and only house the `render` event here.

File: packages/teleport/src/DesktopSession/DesktopSession.tsx
Comment (ibeckermayer) [reply packages/teleport/src/DesktopSession/DesktopSession.tsx:65]: As this is written it's not only associated with the initialization attempt -- `close` can be caused by the server severing the connection at any time.

Let me think about your proposition above (https://github.com/gravitational/webapps/pull/391#discussion_r708007444). Going with that may solve the issue.

File: packages/teleport/src/DesktopSession/TopBar.tsx
Comment (ibeckermayer) [reply packages/teleport/src/DesktopSession/TopBar.tsx:38]: I'm confused -- why does it need to be defined outside of this comp? It relies on `props.attempt`

File: packages/teleport/src/DesktopSession/TopBar.tsx
Comment (ibeckermayer) [reply packages/teleport/src/DesktopSession/TopBar.tsx:148]: > hm, wondering why we couldn't just inline the styles?

We could. Initially I had it that way, but there was enough repetition that I elected for this approach. I was also finding it difficult to edit/read the JSX with so many ternary operator statements in there.

> or we could also pass the attempt down to a styled comp and do our checking for states there?

I had something like that initially but I prefer the `useTheme` hook. The styled comp has so much implicit "magic" in comparison.

File: packages/teleport/src/DesktopSession/useDesktopSession.tsx
Comment (ibeckermayer) [reply packages/teleport/src/DesktopSession/useDesktopSession.tsx]: Great idea, that worked.

File: packages/teleport/src/DesktopSession/useDesktopSession.tsx
Comment (alex-kovoy) [inline packages/teleport/src/DesktopSession/useDesktopSession.tsx:99]: looks complicated, I wonder why did not you like the previous version where tdpClient was close to `canvas` element?

File: packages/teleport/src/DesktopSession/ActionMenu.tsx
Comment (kimlisa) [reply packages/teleport/src/DesktopSession/ActionMenu.tsx]: > Why can I pass color as an object field, but need to pass fontSize in the style object?

Because of styled systems, if you dig into `MenuIcon` you'll see this: https://github.com/gravitational/webapps/blob/master/packages/design/src/ButtonIcon/ButtonIcon.jsx#L93, I think in order for `fontSize` to be recognized as a prop, we need to also include `{typography}`: https://styled-system.com/api/#typography

> How/where does the string 'text.secondary' get interpreted to mean props.theme.colors.text.secondary? Or is that a misunderstanding of what the StyledThemeProvider is passing down to child objects?

From my understanding, the styled-components `ThemeProvider` allows styled components to access the theme object: https://styled-components.com/docs/advanced#theming. If you dig into `MenuIcon`, you'll see we pass these styling props to a styled component: https://github.com/gravitational/webapps/blob/master/packages/design/src/ButtonIcon/ButtonIcon.jsx#L61

File: packages/teleport/src/DesktopSession/TopBar.tsx
Comment (kimlisa) [reply packages/teleport/src/DesktopSession/TopBar.tsx:30]: ignore, it was tied to https://github.com/gravitational/webapps/pull/391#discussion_r708421885

File: packages/teleport/src/DesktopSession/DesktopSession.tsx
Comment (kimlisa) [inline packages/teleport/src/DesktopSession/DesktopSession.tsx:35]: don't think this comment is accurate anymore

File: packages/teleport/src/DesktopSession/useDesktopSession.tsx
Comment (kimlisa) [inline packages/teleport/src/DesktopSession/useDesktopSession.tsx:25]: feels a bit weird to import something from `DesktopList`, maybe have `Desktops/index.ts` re-export it?

File: packages/teleport/src/DesktopSession/TopBar.tsx
Comment (kimlisa) [reply packages/teleport/src/DesktopSession/TopBar.tsx:135]: i think using styled components is the better approach but we can also get a second opinion from @alex-kovoy

File: packages/teleport/src/DesktopSession/useDesktopSession.tsx
Comment (ibeckermayer) [reply packages/teleport/src/DesktopSession/useDesktopSession.tsx:99]: I think you're referring to [this version here](https://github.com/gravitational/webapps/blob/222768dd9bb95843c1950032965948aa79a45c22/packages/teleport/src/DesktopSession/DesktopSession.tsx#L44-L88). The original impetus for moving away from that design was Lisa's comment [here](https://github.com/gravitational/webapps/pull/391#discussion_r707970871), where she suggested keeping state management encapsulated in the `useDesktopSession` hook (my sense is this is standard practice in the codebase with infrequent exceptions). 

This implementation does that, while at the same time implementing Lisa's other suggestion [here](https://github.com/gravitational/webapps/pull/391#discussion_r708007444). With this, we only need to track a single `DesktopSessionAttempt` for the entire connection state, rather than tracking both `wsAttempt` and `fetchDesktopAttempt`, and use [this gross and complex logic to synchronize them](https://github.com/gravitational/webapps/blob/222768dd9bb95843c1950032965948aa79a45c22/packages/teleport/src/DesktopSession/useDesktopSession.tsx#L81-L99).

I agree though, that this is complicated. Another approach lies with making `tdpClient.connect()` [return a `Promise`](https://github.com/gravitational/webapps/blob/2790402525dcfb43a73eb7789004ae151e752c70/packages/teleport/src/lib/tdp/client.ts#L35-L69). Then, this big disgusting thing can just be
```typescript
  useEffect(() => {
    Promise.all([
      ctx.desktopService.fetchDesktops(clusterId),
      tdpClient.connect(),
    ])
      .then(vals => {
        makeUserHost(vals[0]);
        setAttempt({ status: 'success' });
      })
      .catch(err => {
        setAttempt({ status: 'failed', statusText: err.message });
      });
  }, [clusterId, username, desktopId]);
```

(I had this in a previous commit, and then created this mess based on what I now see as a mistaken premise that it was important to keep the `tdpClient` an `EventEmitter` exclusively, and not confuse a user by giving them the option to deal with a `Promise` _or_ an `'open'` Event. That doesn't hold up to scrutiny though, since nobody is confused...)

In that case, we have the option of either adding these event listeners here or in `DesktopSession` next to the `canvas`. My argument for keeping them in here is that by doing so, `tdpClient` events that create side effects by changing state are encapsulated in the hook, alongside any api calls (currently `ctx.desktopService.fetchDesktops(clusterId)`) and the initial websocket connection which do the same. Only the `'render`' event listener need be in `DesktopSession`, distinguished by the fact that it doesn't change the component's state and requires the `canvas` to draw to.

Comment (alex-kovoy) [reply packages/teleport/src/DesktopSession/useDesktopSession.tsx:99]: I think we are dealing with 2 states: one is the session and another one is the canvas. Sometimes it's OK to lift a state up and sometimes it does not. In this case I would keep it closer to the `canvas` as it makes it easier to reason about (because tdpClient _updates_ canvas, canvas is the state and tdpClient is an API).

**Comment by russjones** [comment]
@ibeckermayer Once you have something kind of working, can you do a quick demo for @klizhentas?

File: packages/teleport/src/DesktopSession/useDesktopSession.tsx
Comment (alex-kovoy) [inline packages/teleport/src/DesktopSession/useDesktopSession.tsx:53]: lets move it to the top

File: packages/teleport/src/DesktopSession/DesktopSession.tsx
Comment (alex-kovoy) [inline packages/teleport/src/DesktopSession/DesktopSession.tsx:90]: consider simplifying it a bit if you have time

**Review by alex-kovoy — APPROVED** [review]

**Pull request merged**
